### PR TITLE
refactor: rewrite in node for less tool clutter

### PR DESCRIPTION
**Pull request by m90**
As discussed, the current state using bash, Node and Python for a pretty small piece of software might be detrimental for long term maintainability of this thing.

### CODE REVIEW

File: docker-entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
+
+set -eu
+set -o pipefail
+
+trap finish EXIT INT TERM
```
Comment (m90) [inline docker-entrypoint.sh:6]: This could also be moved into the Node script, but I felt the `trap` would be more suitable

File: Dockerfile [hunk 2]
```diff
@@ -1,9 +1,9 @@
...
+RUN apk add --no-cache curl=8.9.0-r0
 
-COPY --chmod=755 ./transferbot.sh /usr/bin/transferbot
-COPY --chmod=755 ./mangle_data.py /usr/bin/mangle_data
+COPY . /pkg
+RUN npm i -g /pkg
+COPY --chmod=755 docker-entrypoint.sh /usr/bin/docker-entrypoint
 
-ENTRYPOINT ["transferbot"]
+ENTRYPOINT ["docker-entrypoint"]
```
Comment (m90) [inline Dockerfile:9]: This likely breaks how the image is called from api, but I have the feeling it's not done correctly there anyways (explictly relying on `transferbot` being available.

File: bin/cmd.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      console.log(result)
+    }
+  })
+  .catch((err) => {
+    console.error(err)
+    process.exitCode = 1
+  })
+  .then(async () => {
+    if (process.env.CALLBACK_ON_SUCCESS && process.exitCode === 0) {
+      await execCmd(process.env.CALLBACK_ON_SUCCESS)
```
Comment (m90) [inline bin/cmd.js:40]: Running the callbacks from in here now means we don't catch SIGINT, but I guess that's ok for our use case?

File: bin/cmd.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    }
+    if (process.env.CALLBACK_ON_FAILURE && process.exitCode !== 0) {
+      await execCmd(process.env.CALLBACK_ON_FAILURE)
+    }
+  })
+  .catch((err) => {
```
Comment (m90) [inline bin/cmd.js:70]: Interesting question: if the transfer itself worked, but the callback failed, the script will now still exit 0. Is this what we want? I'm unsure.

File: bin/cmd.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    if (result) {
+      console.log(result)
+    }
+  })
+  .catch((err) => {
+    console.error(err)
+    process.exitCode = 1
+  })
+  .then(() => {
+    if (process.env.CALLBACK_ON_SUCCESS && process.exitCode === 0) {
```
Comment (m90) [inline bin/cmd.js:63]: Running the callbacks from in here now means we don't catch SIGINT, but I guess that's ok for our use case?

File: bin/cmd.js (hunk 4 above)
Comment (tarrow) [reply bin/cmd.js:70]: After trying to think through the steps that would happen in this case I believe it's probably no worse than returning `1`.

In either case because of how we specified the spec the job won't retry.

This means that it may be marginally less clear to the debugging operator but on the other hand the "main work" done by the job did succeed.

File: bin/cmd.js (hunk 5 above)
Comment (tarrow) [reply bin/cmd.js:63]: I'm not sure if this might leave us with a larger (but still small) chance of wiki's being left in a partially imported state. In any case I don't think k8s is likely to SIGINT but rather SIGTERM. If we wanted to be super belt and braces we could trap this and then CALLBACK_ON_FAILURE

File: lib/repo.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+      .then(r => r.json())
+      .then(body => Object.keys(body.query.wbcontentlanguages))
+  }
+
+  createEntities (...entities) {
+    if (!this.edit) {
+      return Promise.reject(new Error('Cannot edit a read only instance.'))
+    }
+    return Promise.all(entities.map(async entity => {
+      return this.edit.entity.create(entity)
```
Comment (tarrow) [inline lib/repo.js]: Does this stop the whole import if one entity fails to be created? I think I've convinced myself that we would rather have this continue even if one entity fails

File: bin/cmd.js (hunk 5 above)
Comment (m90) [reply bin/cmd.js:63]: There is a cleanup Laravel job that sweeps stalled imports every hour or so.

We could leave the trap in bash as is if we wanted to, but then we'd have a bit of mish-mash still. Which is what we are trying to leave behind.

File: lib/repo.js (hunk 6 above)
Comment (m90) [reply lib/repo.js]: This rejects when the first promise in the given array rejects, so it's not entirely predictable how much work is done in such a case. We should be able to use `allSettled` though and get what you describe. Question though is: do we consider such a case a failure still and exit 1?

Comment (m90) [reply lib/repo.js]: Added this in https://github.com/wbstack/transferbot/pull/7/commits/9da79ce0b13fe806410fca476fcf5b1b7c858cee

File: bin/cmd.js (hunk 5 above)
Comment (m90) [reply bin/cmd.js:63]: Added some Node based trapping in https://github.com/wbstack/transferbot/pull/7/commits/214be4d9973acb7e478021fccddbcd259fe27251 that seems to work when double CRL+Cing the thing running in Docker

```
➜  transferbot git:(fr/nodeify) ✗ docker run -e TARGET_WIKI_OAUTH_CONSUMER_TOKEN="xxx" -e TARGET_WIKI_OAUTH_CONSUMER_SECRET="xxx" -e TARGET_WIKI_OAUTH_ACCESS_TOKEN="xxx" -e TARGET_WIKI_OAUTH_ACCESS_SECRET="xxx" -e CALLBACK_ON_FAILURE="echo 'failing'" --rm 4ab54cf0400349e5e0afbf06e1965046c1b7cf0f8890c505afd8b733efac3a07 https://www.wikidata.org https://updatewiki.wikibase.dev Q107819273
^CAn error occurred executing the command:
Error: Received SIGINT, program will exit before finishing.
    at process.<anonymous> (file:///pkg/lib/signals.js:5:16)
    at process.emit (node:events:519:28)
failing
^C%
```

Single Ctrl+C would have Node waiting for its event loop to become empty and would trigger both the transfer as well as the failure callback. I would think this is a problem we'd also have in bash.

File: bin/cmd.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env node
+
+import WikibaseRepo from './../lib/repo.js'
+import { pickLanguages, pickKeys } from './../lib/util.js'
+import { execCmd } from './../lib/exec.js'
+import { trap } from './../lib/signals.js'
+
+Promise.race([
```
Comment (tarrow) [inline bin/cmd.js:10]: praise: I needed to quickly look up how this fancy thing worked and that's very cool. Looks suitably tidy!

File: bin/cmd.js (hunk 5 above)
Comment (tarrow) [reply bin/cmd.js:63]: I ended up needing to do a load of reading the docs about `process` and this then inspired me to ask: "should we consider using `beforeExit` for all this stuff?" see https://nodejs.org/api/process.html#event-beforeexit

File: lib/repo.js (hunk 6 above)
Comment (tarrow) [reply lib/repo.js]: nice; part of my reasoning is that it means rerunning the job will be less painful in future (although the rerun will also error).

**Comment by tarrow** [comment]
I'm not finished reviewing this; I really wanted to "just try it" but need to look some more tomorrow about how to best actually do this.

We're you developing by building the image and then adding it into minikube? Or using skaffold? Or building it locally and running locally after exfiltrating credentials from your local development environment? Or even something else?

**Comment by m90** [comment]
> Or building it locally and running locally after exfiltrating credentials from your local development environment?

This is how I tested it, yes.

File: bin/cmd.js (hunk 5 above)
Comment (m90) [reply bin/cmd.js:63]: So this is a bit ugly. When using `beforeExit`, a forceful exit (i.e. Ctrl+C) or an unhandled exception will not be considered:

> The 'beforeExit' event is not emitted for conditions causing explicit termination, such as calling [process.exit()](https://nodejs.org/api/process.html#processexitcode) or uncaught exceptions.

In theory `on('exit')` would help us, however this (as opposed to `beforeExit`) only allows for synchronous code to be run in the callback:

> Listener functions must only perform synchronous operations. The Node.js process will exit immediately after calling the 'exit' event listeners causing any additional work still queued in the event loop to be abandoned.

 meaning we can't shell out to the command the user is passing.

Maybe we're already at a point where this is "as good as it gets" if we want to keep this in Node.

Comment (m90) [reply bin/cmd.js:63]: Maybe it's good to remind ourselves to not overcomplicate this and aim for a solution that ensures the given commands will ALWAYS run as expected.

Comment (m90) [reply bin/cmd.js:63]: I implemented some cancellation mechanism in 49f77f1acf9638aba49752eb1117cb6339ead55c

This works as intended now, but I am not 100% sure we should use it as it makes the code kind of hard to follow. This is mostly because the wikibase-x librarys do not expose the fetch they are using, so we cannot use the plain `signal` but have to add some custom wrapper, when `fetch` calls can use the signal as intended. I'm not really sure if the extra complexity justifies the outcome.

**Comment by m90** [comment]
Looking at this in its current state I have to admit I would lean towards not merging this, as it is now FAR more complex than what we had, even if it's just using a single language. To me, this is a bad tradeoff. But since I am not the person who'll be maintaining this, I'll leave the decision up to you.

**Comment by m90** [comment]
As discussed, this is not what we want as it makes things more complex instead of giving us simplicity.
